// @remotion/google-fonts/dist/cjs/BJCree.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadFont = exports.fontFamily = exports.getInfo = void 0;
const base_1 = require("./base");
const getInfo = () => ({
    fontFamily: 'BJ Cree',
    importName: 'BJCree',
    version: 'v1',
    url: 'https://fonts.googleapis.com/css2?family=BJ+Cree:ital,wght@0,400;0,500;0,600;0,700',
    unicodeRanges: {
        'canadian-aboriginal': 'U+02C7, U+02D8-02D9, U+02DB, U+0307, U+1400-167F, U+18B0-18F5, U+25CC, U+11AB0-11ABF',
        latin: 'U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA, U+02DC, U+0304, U+0308, U+0329, U+2000-206F, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF, U+FFFD',
    },
    fonts: {
        normal: {
            '400': {
                'canadian-aboriginal': 'https://fonts.gstatic.com/s/bjcree/v1/8QIJdijLoYvJ7b7biKAEqojk.woff2',
                latin: 'https://fonts.gstatic.com/s/bjcree/v1/8QIJdijLoYvJ7b7biI8Eqg.woff2',
            },
            '500': {
                'canadian-aboriginal': 'https://fonts.gstatic.com/s/bjcree/v1/8QIMdijLoYvJ7b7bgHwnv5jFHdVy.woff2',
                latin: 'https://fonts.gstatic.com/s/bjcree/v1/8QIMdijLoYvJ7b7bgHwnv7fFHQ.woff2',
            },
            '600': {
                'canadian-aboriginal': 'https://fonts.gstatic.com/s/bjcree/v1/8QIMdijLoYvJ7b7bgFAgv5jFHdVy.woff2',
                latin: 'https://fonts.gstatic.com/s/bjcree/v1/8QIMdijLoYvJ7b7bgFAgv7fFHQ.woff2',
            },
            '700': {
                'canadian-aboriginal': 'https://fonts.gstatic.com/s/bjcree/v1/8QIMdijLoYvJ7b7bgDQhv5jFHdVy.woff2',
                latin: 'https://fonts.gstatic.com/s/bjcree/v1/8QIMdijLoYvJ7b7bgDQhv7fFHQ.woff2',
            },
        },
    },
    subsets: ['canadian-aboriginal', 'latin'],
});
exports.getInfo = getInfo;
exports.fontFamily = 'BJ Cree';
const loadFont = (style, options) => {
    return (0, base_1.loadFonts)((0, exports.getInfo)(), style, options);
};
exports.loadFont = loadFont;
